enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,640 @@
+PerformanceProfile API to include settings for workload partitioning.
+
+When installation is done and workload partitioning is set then from that point
+on the `kubelet` and `crio` only need to be configured with the desired CPU set
+to use. We currently express this to customers via the `reserved` CPU set as
+part of the performance profile api.
+
+We want to add a new `workloads` field to the `cpu` field that contains a list
+of enums for defining which workloads to pin. This should allow us to expand
+this in the future if desired, but in this enhancement we will only support
+`Infrastructure` which defines all of the Openshift workloads.
+
+```yaml
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: openshift-node-workload-partitioning-custom
+spec:
+  cpu:
+    isolated: 2-3
+    reserved: 0,1
+    # New addition
+    workloads:
+      - Infrastructure
+```
+
+To support upgrades and maintain better signaling for the cluster, the
+Performance Profile Controller will also inspect the Nodes to update a global
+identifier at start up. We will only update the identifier to `AllNodes` if we
+are running in Single Node and our Node has the capacity resource
+(`management.workload.openshift.io/cores`) for our 4.11 -> 4.12 upgrades. This
+should not be needed after 4.12 for all clusters. This should have no baring on
+4.11 HA/3NC clusters as this feature will not be back ported.
+
+### Workflow Description
+
+The end user will be expected to provide the default machine configs to turn on
+the feature for the whole cluster. As well as provide a `PerformanceProfile`
+manifest that describes their desired `isolated` and `reserved` CPUSet and the
+`Infrastructure` enum provided to the list in the `workloads` enum list.
+
+**High level sequence diagram:**
+
+Install Time Sequence
+
+```mermaid
+sequenceDiagram
+Alice->>Installer: Generate Manifests
+Installer->>Alice: Create Dir with Cluster Manifests
+loop Apply
+    Alice->>Alice: Add Custom Machine Configs
+    Alice->>Alice: Add PerformanceProfile Manifest
+    Alice->>Alice: Alter Infrastructure Status
+end
+Alice->>Installer: Create Cluster
+Installer-->>APIServer: Apply Machine Configs
+APIServer-->>Installer: Applied!
+APIServer-->>MCO: Machine Manifests
+MCO-->>Node: Configure and Restart Node
+loop Apply
+    Node->>Node: Set Kubelet Config
+    Node->>Node: Set CRI- Config
+    Node->>Node: Kubelet Advertises Cores
+end
+Node-->>MCO: Finished Restart
+loop Generate
+    NTO->>NTO: Machine Config from PerformanceProfile
+end
+NTO-->>MCO: Apply Machine Config
+MCO-->>Node: Configure and Restart Node
+loop Apply
+    Node->>Node: Set Kubelet Config
+    Node->>Node: Set CRI- Config
+end
+Node-->>MCO: Finished Restart
+```
+
+- **Alice** is a human user who creates an Openshift cluster.
+- **Installer** is assisted installer that applies the user manifest.
+- **MCO** is the machine config operator.
+- **Node** is the kubernetes node.
+
```

Comment:
Would this process work with the agent installer?

/cc @zaneb

Reply:
Creating an extra manifest like a MachineConfig is easy.
_Editing_ one of the manifests created by the installer is not something that is exposed by the agent installer at the moment, so that part could be challenging.

Reply:
The approach here has been altered, after some discussion this has been changed to be an install-config.yaml option